### DISCUSSION TITLE
Azure Appservice

### DISCUSSION THREAD

**Discussion by avialonn** [hardkoded/puppeteer-sharp#2423] [Q&A]
Hi All,
Is it possible to run this on azure AppService (Windows)?
I'm getting the following error with no success:

An error occurred trying to start process 'C:\home\site\wwwroot\Chrome\Win64-121.0.6167.85\chrome-win64\chrome.exe' with working directory 'C:\home\site\wwwroot'. The application has failed to start because its side-by-side configuration is incorrect. Please see the application event log or use the command-line sxstrace.exe tool for more detail. line: 208

Thanks!
AA

**Comment by kblok** [comment]
Docker images proved to be the best way to run puppeteer in the cloud.

**Reply by avialonn** [reply]
Got it! I will give it a try.
Thanks

**Comment by sashiyama** [comment]
I've created a docker image to run a proxy server that provide WebSocket endpoint to use Chromium headless browser from your application. You can run it on your local environment or server such as render.com, Azure VMs, and so on even though your application is running on Azure App Service that does not allow apps to run headless browsers.

https://github.com/sashiyama/browserless
```csharp
var options = new ConnectOptions
{
    BrowserWSEndpoint = "ws://localhost/chromium?token=xxx"
};

return await Puppeteer.ConnectAsync(options);
```

**Reply by John-PS** [reply]
we had a web app service running this, but as soon as we upgraded to .net core 8 (and updated the old puppeteer Sharp) it is essentially broken as just calling the Page.GotoAsync makes it throw the error regarding the detached Frame. I've found a lot of notes about that and it sounds like it just doesnt work anymore:

PuppeteerException: Navigating frame was detached
PuppeteerSharp.Helpers.TaskHelper.WithTimeout(Task task, Func<Task> timeoutAction, TimeSpan timeout, CancellationToken cancellationToken) in TaskHelper.cs, line 101

NavigationException: Navigating frame was detached
PuppeteerSharp.Cdp.CdpFrame.GoToAsync(string url, NavigationOptions options) in CdpFrame.cs, line 108

**Reply by sashiyama** [reply]
@John-PS  Thank you for letting me know. It seems like my node library and Puppeteer Sharp don't have compatibility anymore since you upgraded the library. I think that since my library is just wrapping Puppeteer with WebSocket server, if my library and Puppeteer sharp use same version of Puppeteer, it should work. Have you rebuild my library docker image since you upgraded Puppeteer Sharp? If you haven't, could you do that to update puppeteer in my library running on your server.

**Reply by John-PS** [reply]
I should clarify - I meant that we were using Puppeteer Sharp and it stopped working when we moved to .net core 8; we aren't using a docker image  as the project we inherited is using a slightly different setup.

**Comment by DevNamedZed** [comment]
We were able to get working on windows azure app service. Issue with sxs was due to path . use HOME_EXPANDED env var

```
private async Task<string> DownloadAsync()
{
    var homeExpanded = Environment.GetEnvironmentVariable("HOME_EXPANDED")
        ?? Environment.GetEnvironmentVariable("HOME")
        ?? Path.GetTempPath();

    var downloadPath = Path.Combine(homeExpanded, "Chrome");

    _logger.LogInformation("Downloading Chromium to {DownloadPath}", downloadPath);

    var browserFetcher = new BrowserFetcher(new BrowserFetcherOptions
    {
        Path = downloadPath
    });

    var installedBrowser = await browserFetcher.DownloadAsync();

    _logger.LogInformation("Chromium downloaded to {ExecutablePath}", installedBrowser.GetExecutablePath());

    return installedBrowser.GetExecutablePath();
}
```

**Reply by 93Omar** [reply]
> We were able to get working on windows azure app service. Issue with sxs was due to path . use HOME_EXPANDED env var
> 
> ```
> private async Task<string> DownloadAsync()
> {
>     var homeExpanded = Environment.GetEnvironmentVariable("HOME_EXPANDED")
>         ?? Environment.GetEnvironmentVariable("HOME")
>         ?? Path.GetTempPath();
> 
>     var downloadPath = Path.Combine(homeExpanded, "Chrome");
> 
>     _logger.LogInformation("Downloading Chromium to {DownloadPath}", downloadPath);
> 
>     var browserFetcher = new BrowserFetcher(new BrowserFetcherOptions
>     {
>         Path = downloadPath
>     });
> 
>     var installedBrowser = await browserFetcher.DownloadAsync();
> 
>     _logger.LogInformation("Chromium downloaded to {ExecutablePath}", installedBrowser.GetExecutablePath());
> 
>     return installedBrowser.GetExecutablePath();
> }
> ```

Can you provide a full sample code? I tried your code but I'm getting this error:
```
PuppeteerSharp.ProcessException: Failed to launch browser! 
   at PuppeteerSharp.States.ProcessStartingState.StartCoreAsync(LauncherBase p) in /home/runner/work/puppeteer-sharp/puppeteer-sharp/lib/PuppeteerSharp/States/ProcessStartingState.cs:line 83
   at PuppeteerSharp.States.ProcessStartingState.StartCoreAsync(LauncherBase p) in /home/runner/work/puppeteer-sharp/puppeteer-sharp/lib/PuppeteerSharp/States/ProcessStartingState.cs:line 89
   at PuppeteerSharp.Launcher.LaunchAsync(LaunchOptions options) in /home/runner/work/puppeteer-sharp/puppeteer-sharp/lib/PuppeteerSharp/Launcher.cs:line 77
   at PuppeteerSharp.Launcher.LaunchAsync(LaunchOptions options) in /home/runner/work/puppeteer-sharp/puppeteer-sharp/lib/PuppeteerSharp/Launcher.cs:line 110
```
I'm using .NET 8 with PuppeteerSharp v.20.2.6

**Reply by malighieri** [reply]
I did run into the same problem. It's important to notice that the error now is different from the original one.
It went from "An error occurred trying to start process" to "Failed to launch browser".
At the end, the browser did't start the same, but it seems the problem changed.

**Reply by malighieri** [reply]
In my research, I discovered that the hermetic environment of Azure Web App (or Azure Functions) can still limit Chrome's startup. To try and work around these limitations, I used several arguments in the startup process (listed below), and even then, the "Failed to start browser!" error persisted.

--no-sandbox: Disables Chrome's security sandbox, allowing the browser to run without process isolation restrictions. Necessary in restricted environments such as Docker, CI pipelines, or Azure App Service.

--disable-setuid-sandbox: Disables the setuid-based sandbox mechanism, primarily used on Linux systems. Usually used in conjunction with --no-sandbox when the environment does not support setuid sandboxing.

--disable-dev-shm-usage: Prevents Chrome from using the /dev/shm shared memory directory, forcing it to use /tmp. Useful in containers or environments with limited shared memory (e.g., Docker, Kubernetes). ... --disable-gpu: Disables GPU hardware acceleration. Useful on servers without GPU support or in environments without a graphical interface.

--allow-file-access-from-files: Allows pages loaded via file:// to access other local files. Necessary when local HTML files load CSS, JavaScript, images, or other local resources.

--disable-web-security: Disables various web security protections, including Same-Origin Policy and CORS restrictions. Used in automation scenarios where pages need to access resources from different origins without browser restrictions.

--disable-crash-reporter: Disables Chrome's crash reporting system (Crashpad), which normally sends crash reports. Useful in restricted environments where the crash reporting process cannot be initiated.

--disable-breakpad: Disables Breakpad, Chromium's internal crash handling mechanism. Helps reduce helper processes and avoid crash handling issues in restricted environments.

--no-zygote: Disables the zygote process, which Chrome normally uses to efficiently start new browser processes. Useful in environments that restrict the creation of child processes.

--single-process: Forces Chrome to run all components in a single process, instead of the usual multi-process architecture. Sometimes used in environments with strict process limits or low memory, although it may reduce stability and isolation.
